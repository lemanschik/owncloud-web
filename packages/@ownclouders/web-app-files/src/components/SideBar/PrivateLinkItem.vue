<template>
  <oc-button
    v-if="isClipboardCopySupported"
    v-oc-tooltip="buttonLabel"
    appearance="raw"
    :aria-label="buttonLabel"
    class="oc-files-private-link-copy-url"
    :variation="copied ? 'success' : 'passive'"
    @click="copyLinkToClipboard"
  >
    <span v-text="buttonText" />
    <oc-icon
      v-if="copied"
      key="oc-copy-to-clipboard-copied"
      name="checkbox-circle"
      class="_clipboard-success-animation"
    />
    <oc-icon v-else key="oc-copy-to-clipboard-copy" name="clipboard" />
  </oc-button>
</template>

<script lang="ts">
import { computed, defineComponent, inject, unref } from 'vue'
import { useClipboard } from '@vueuse/core'
import { Resource } from 'web-client'
import { useStore } from 'web-pkg/src'
import { useGettext } from 'vue3-gettext'

export default defineComponent({
  name: 'PrivateLinkItem',
  setup() {
    const { $gettext } = useGettext()
    const store = useStore<any>()
    const resource = inject<Resource>('resource')
    const privateLink = computed(() => unref(resource))

    const {
      copy,
      copied,
      isSupported: isClipboardCopySupported
    } = useClipboard({ legacy: true, copiedDuring: 550 })

    const copyLinkToClipboard = () => {
      copy(unref(privateLink).privateLink)
      store.dispatch('showMessage', {
        title: $gettext('Private link copied'),
        desc: $gettext('The private link has been copied to your clipboard.')
      })
    }

    return {
      copied,
      copyLinkToClipboard,
      isClipboardCopySupported
    }
  },
  computed: {
    buttonText() {
      return this.$gettext('Private link')
    },
    buttonLabel() {
      return this.$gettext('Copy private link to clipboard')
    }
  }
})
</script>
