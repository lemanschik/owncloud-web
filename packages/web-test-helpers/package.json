{
  "name": "web-test-helpers",
  "private": true,
  "main": "src/index.ts",
  "peerDependencies": {
    "@casl/ability": "^6.3.3",
    "@casl/vue": "^2.2.1",
    "axios": "0.27.2",
    "design-system": "@ownclouders/design-system",
    "vue3-gettext": "^2.3.3",
    "vue-router": "4.1.6",
    "vuex": "4.1.0"
  }
}
