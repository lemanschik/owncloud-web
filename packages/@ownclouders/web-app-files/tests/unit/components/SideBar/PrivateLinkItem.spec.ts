import { mock } from 'jest-mock-extended'
import { Resource } from 'web-client'
import { createStore, defaultPlugins, mount, defaultStoreMockOptions } from 'web-test-helpers'
import PrivateLinkItem from 'web-app-files/src/components/SideBar/PrivateLinkItem.vue'

jest.useFakeTimers()

const folder = mock<Resource>({
  type: 'folder',
  ownerId: 'marie',
  ownerDisplayName: 'Marie',
  mdate: 'Wed, 21 Oct 2015 07:28:00 GMT',
  size: '740',
  name: 'lorem.txt',
  privateLink: 'https://example.com/fake-private-link'
})

describe('PrivateLinkItem', () => {
  it('should render a button', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('upon clicking it should copy the private link to the clipboard button, render a success message and change icon for half a second', async () => {
    Object.assign(window.navigator, {
      clipboard: {
        writeText: jest.fn().mockImplementation(() => Promise.resolve())
      }
    })

    const { wrapper, storeOptions } = getWrapper()
    expect(storeOptions.actions.showMessage).not.toHaveBeenCalled()

    await wrapper.trigger('click')
    expect(wrapper.html()).toMatchSnapshot()
    expect(window.navigator.clipboard.writeText).toHaveBeenCalledWith(folder.privateLink)
    expect(storeOptions.actions.showMessage).toHaveBeenCalledTimes(1)

    jest.advanceTimersByTime(550)

    wrapper.vm.$nextTick(() => {
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})

function getWrapper() {
  const storeOptions = { ...defaultStoreMockOptions }
  storeOptions.getters.capabilities.mockImplementation(() => ({ files: { privateLinks: true } }))
  const store = createStore(storeOptions)
  return {
    storeOptions,
    wrapper: mount(PrivateLinkItem, {
      global: {
        plugins: [...defaultPlugins(), store],
        provide: {
          resource: folder
        }
      }
    })
  }
}
