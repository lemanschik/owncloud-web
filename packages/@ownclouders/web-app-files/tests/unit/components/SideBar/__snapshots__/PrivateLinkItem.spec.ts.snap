// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateLinkItem should render a button 1`] = `
<button aria-label="Copy private link to clipboard" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-files-private-link-copy-url" type="button">
  <!-- @slot Content of the button -->
  <span>Private link</span>
  <span class="oc-icon oc-icon-m oc-icon-passive">
    <!---->
  </span>
</button>
`;

exports[`PrivateLinkItem upon clicking it should copy the private link to the clipboard button, render a success message and change icon for half a second 1`] = `
<button aria-label="Copy private link to clipboard" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-success oc-button-success-raw oc-files-private-link-copy-url" type="button">
  <!-- @slot Content of the button -->
  <span>Private link</span>
  <span class="oc-icon oc-icon-m oc-icon-passive _clipboard-success-animation">
    <!---->
  </span>
</button>
`;

exports[`PrivateLinkItem upon clicking it should copy the private link to the clipboard button, render a success message and change icon for half a second 2`] = `
<button aria-label="Copy private link to clipboard" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-files-private-link-copy-url" type="button">
  <!-- @slot Content of the button -->
  <span>Private link</span>
  <span class="oc-icon oc-icon-m oc-icon-passive">
    <!---->
  </span>
</button>
`;
